### DISCUSSION TITLE
RUN steps in `.DockerBuild` should behave as `WithExec`

### DISCUSSION THREAD

**Discussion by jedevc** [dagger/dagger#12387] [maintainers]
:bulb: Inspired @shykes's comment in [discord](https://discord.com/channels/707636530424053791/1122940573767184546/1171623607194099792).

There's a few extra steps that `WithExec` takes, that currently `docker run` doesn't, mostly a lot of setting `_DAGGER` internal env vars.

Importantly though, these enable a few useful features :tada:

- Services https://github.com/dagger/dagger/blob/38c168837bcbb4c33bce51306092b9ce1a1e968c/core/container.go#L1082-L1087
- Secret scrubbing (even though secrets can be passed through :scream:) https://github.com/dagger/dagger/blob/38c168837bcbb4c33bce51306092b9ce1a1e968c/core/container.go#L1159-L1169

  Currently, secrets only get stripped from the result of the entrypoint, not from any of the build steps, see the test [here](https://github.com/dagger/dagger/blob/bdaefc4d9769c949618b70e521aa4094e58d50b2/core/integration/container_test.go#L219-L236).

`RUN` called in a `Dockerfile` doesn't pass through any of this, so none of these features work from there :cry: 

If they did, this would allow the following pattern to work:

```go
func (m *Playground) Build(svc *Service) *Container {
	dir := dag.Host().Directory(".")
	return dag.Container().WithServiceBinding("server", svc).Build(dir)
}
```

It's quite a common issue that there's no way of connecting services (like databases) into dockerfile build steps - it would be really nice to actually solve this once and for all.

---

This isn't just for `RUN` steps though - git/http sources in `Dockerfile`s should use services provided by the container.

Doing this is definitely a bit painful though :cry: We'd need to mutate `ExecOp`s/`Source`s in the llb that the frontend produces directly in the solver, or switch to calling the frontend manually and mutating the llb from within the call to `DockerBuild`. From there, we could set the right env vars, replace the git/http sources with our own, etc. Maybe my understanding's not quite right though, and this is easier than I'm thinking?

cc @sipsma @vito

**Comment by jedevc** [comment]
There's also a few other cases where it's possible to leak secrets in logs, specifically in command names (which appear in progrock)

Not *quite* sure how we would get this out actually, ideally we could have a progrock adapter for the secret censor (though we'd need to avoid processing the already processed logs from ExecOps).

**Comment by vaniobegic** [comment]
Not only is this +1, but further it is currently not obvious that building container images usign the build commands shares anything with Dagger engine. The context within which the steps from Dockerfile are finally executed is not clearly documented.
